### ISSUE TITLE
[Discussion] Prepare views in advance of navigation

### ISSUE THREAD

**Issue by ChaseFlorell** [PrismLibrary/Prism#1475]
I have a thought about performance in navigation with a proposed API addition.

I’m finding that some heavy pages take a long time to navigate to (especially on Android). This is entirely due to how Xamarin is plumbed, and has a lot to do with the Binding engine. 

In a talk that Jason Smith gave at a past Evolve, he said that pages can be constructed on a background thread and navigated to as long as it’s marshalled back to the UI thread. With this in mind, I wonder if an API addition as follows would be something you’d consider.

```csharp
NavigationItem _staged;

public override void OnNavigatedTo(NavigationParameters parameters)
{
    // get everything ready to navigate to
    _staged = await _navigationService.Prepare("path/to/navigate"); 
}

public void NavigateToSomePage()
{
    var myParameters = new NavigationParameters(); // hydrate, whatever
    await _navigationService.Navigate(_staged, myParameters);
}
```

Doing this, Prism can setup the page and the BindingContext in advance, and `Navigate(NavigationItem navigationItem, NavigationParameters parameters)` would simply invoke the actual navigation.

The API change would be to add an overload for `Navigate` with the above signature, and add a new method called `Prepare` as well as a `NavigationItem` (name TBD).

Thoughts?

I thought about making an extension to do this, but I’m not sure I can hook into Prism’s NavigationService and get the constructed bits separate from a navigation step.

**Comment by AmrAlSayed0** [comment]
Would it be useful to have a(n) Unprepare/Destroy/Cancel/Release method to complement the functionality of the Prepare method?<br/>
The scenario I have in mind is this. Say there is a page called `MainPage` that will either go to one of 2 pages `ViewA` or `ViewB.` When you land on `MainPage,` you will prepare both `ViewA` and `ViewB` and save them as fields. When the user interaction is decided you would release/destroy the page that will not be used and navigate to the other one. Another case is if the user just goes back on the current page and you will need to destroy/release the prepared pages.
<br/>
<br/>
I have a feeling I am wrong about something here. Like is there even a need for this method? Would just dereferencing the `NavigationItem` be enough for it to be garbage collected?

**Comment by dansiegel** [comment]
As discussed offline. There are a number of issues here that I'm not sure we could ever resolve to make this a useful API.

- The goal of the API is to reduce initialization time 
- Initialization is greatly slowed by inflating the Renderer which cannot be done ahead of time
- Initialization is also greatly slowed by Binding Heavy Pages where the Binding Engine bogs down.
- Initialization requires particularly on Binding Heavy Pages tends to require an initialization context that is typically passed via NavigationParameters

Since a Navigation call goes through the following flow:

- CanNavigate (Source)
- OnNavigatingTo (Destination)
- OnNavigatedFrom (Source)
- ONNavigatedTo (Destination)

This scenario becomes further complicated as it would be expected that I may save my context such on ViewA, pass it to ViewB. This solves the initialization issue, except for the fact that when I want to invoke `_navigationService.Navigate` I may have changed state and will now bypass CanNavigate and may be in a state where CanNavigate returns false. 

As another issue, I don't believe that we could really support Deep Linking or if we did it would be under limited scenarios.

Bottom line I think this is likely to introduce more issues than it solves. If I'm missing something here and you want to continue the conversation I'm happy to re-evaluate and reopen the issue if it makes sense.

**Comment by ChaseFlorell** [comment]
I definitely understand the comments above, and upon further contemplation, this API would need the following change.

```csharp
NavigationItem _staged;

public override void OnNavigatedTo(NavigationParameters parameters)
{
    // get everything ready to navigate to
    var myParameters = new NavigationParameters(); // hydrate, whatever
    _staged = await _navigationService.Prepare("path/to/navigate", myParameters); 
}

public void NavigateToSomePage()
{
    // should we allow _staged.NavigationParameters to be changed, or should it be internal?
    await _navigationService.Navigate(_staged); // navigation parameters are already locked in
}
```

Since any time a page and viewmodel is `Prepare`d, we'd know that the user intends on navigating to that page. Therefore we can call `OnNavigatingTo` during the preparation, even if the act of navigated is further down the line.

I'm not sure how deep linking would fall into this, but I _assume_ we can provide all of the details necessary within the `NavigationItem`.

What this ultimately solves is the slow Binding setup from the Xamarin.Forms binding engine... essentially the `page.BindingContext = viewModel;`. I've measured this on binding heavy pages, and have seen it upwards of a couple of seconds. This is a terrible user experience that I'd really like to solve.

**Comment by dansiegel** [comment]
As an example here say you have ViewA with a List of Monkeys. You call Prepare for ViewB which shows the details for a specific Monkey. The problem is that you can't actually prepare ViewB for the Monkey until you've selected the monkey in the list that you want to view the details of. Which means that you don' really gain anything in any way.
